### PR TITLE
Arrow refactoring and performance improvements

### PR DESCRIPTION
**Pull request by fabianmurariu**
### What changes were proposed in this pull request?
Rethink the way we interact with `arrow-core`

### Why are the changes needed?
Redo the ArrowPartition ingestion mechanism using disruptor pattern
Redo the execution by using the arrow-core multi threaded iterator

### Does this PR introduce any user-facing change? If yes is this documented?
No

### How was this patch tested?
Added extra tests for arrow in Java and arrow partition in Scala

### Are there any further changes required?
So many..

### CODE REVIEW

File: arrow-core/src/main/java/com/raphtory/arrowcore/implementation/EdgePartitionManager.java [hunk 1]
```diff
@@ -246,7 +258,7 @@ protected EdgePartition getPartition(int partId) {
      *
      * @return the edge partition
      */
-    protected EdgePartition getPartitionAndLoad(int partId) {
+    public EdgePartition getPartitionAndLoad(int partId) {
         EdgePartition p = _partitions.get(partId);
         if (p == null) {
             if (_partitions.size() >= N_ARROW_PARTITIONS) {
```
Comment (miratepuffin) [inline arrow-core/src/main/java/com/raphtory/arrowcore/implementation/EdgePartitionManager.java:264]: //removeLRUPartition(); 
and 
else {
            //touchPartition(p);
        }

not required below?

File: arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexArrowStore.java [hunk 2]
```diff
@@ -122,12 +128,11 @@ public boolean addVertex(int row, Vertex v) {
         //System.out.println("Added vertex: " + v.getLocalId());
```
Comment (miratepuffin) [inline arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexArrowStore.java:128]: Can be removed?

File: arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java [hunk 3]
```diff
@@ -140,16 +141,16 @@ public int compare(int row1, int row2) {
 
 
 
-    private final int _partitionId;
-    private final VertexPartition _avp;
-    private final VertexPartitionManager _apm;
+    protected final int _partitionId;
+    protected final VertexPartition _avp;
+    protected final VertexPartitionManager _apm;
     //private final VertexSnapshotPartition _snapshot;
```
Comment (miratepuffin) [inline arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java:147]: Remove?

File: arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java [hunk 4]
```diff
@@ -140,16 +141,16 @@ public int compare(int row1, int row2) {
...
-    private boolean _modified = false;
-    private boolean _sorted = false;
+    protected VertexHistoryStore _history;
+    protected VectorSchemaRoot _historyRO;
+    protected ArrowFileReader _historyReader;
+    protected boolean _modified = false;
+    protected boolean _sorted = false;
 
 
     /**
```
Comment (miratepuffin) [inline arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java:156]: //_snapshot = new VertexSnapshotPartition(partitionId, avp); needed below?

Comment (miratepuffin) [reply arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java:156]: Same with //_snapshot.initialize();

Comment (miratepuffin) [reply arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java:156]: And 
        if (updated) {
            //_snapshot.takeSnapshot();
        }

Comment (miratepuffin) [reply arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java:156]: //System.out.println("CLOSING PARTITION: " + partitionId);

File: arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java [hunk 5]
```diff
@@ -472,29 +478,25 @@ protected void sortHistoryTimes() {
         }
 
         _sorted = true;
+
+        //dump();
```
Comment (miratepuffin) [inline arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java:482]: remove

File: arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java [hunk 6]
```diff
@@ -610,48 +605,43 @@ protected void findHistory(VertexHistoryIterator.WindowedVertexHistoryIterator s
...
 
 
+    /**
+     * @return the number of history items in this partition
+     */
     public long getNHistoryItems() {
-        int n = _history._maxRow;
-        return n+1;
+        return _history._maxRow;
     }
```
Comment (miratepuffin) [inline arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java:647]: //System.out.println("TIME at " + row + " : " + creationTime + ", inRange=" + (creationTime>=min && creationTime<=max)); to logger or removed?

File: arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java [hunk 7]
```diff
@@ -954,4 +951,188 @@ public long getVertexMaxHistoryTime(int vertexRowId) {
...
+     * @param searcher the bounded searcher to use
+     *
+     * @return true if the vertex is alive in the window
+     */
+    protected boolean isAliveAt(int vertexRow, long start, long end, BoundedVertexEdgeTimeWindowComparator searcher) {
+        int low = _avp._store._sortedHStart.get(vertexRow);
+        int high = _avp._store._sortedHEnd.get(vertexRow);
+
+        if (low==-1 || high==-1) {
+            System.out.println("Looks like you're searching for history BUT haven't added any for this vertex!");
```
Comment (miratepuffin) [inline arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java:985]: Any reason for not using loggers for things like this?

File: arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java [hunk 8]
```diff
@@ -954,4 +951,188 @@ public long getVertexMaxHistoryTime(int vertexRowId) {
...
+                    _history._times.get(actualRow)>=start &&
+                    _history._times.get(actualRow)<=end;
+        }
+
+        int actualRow = _history._sortedVertexTimeIndices.get(row);
+        return _history._states.get(actualRow)!=0;
+    }
+
+
+    private void dumpBits(int start, int high) {
```
Comment (miratepuffin) [inline arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryPartition.java:1025]: This another debugging function?

File: arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryStore.java [hunk 9]
```diff
@@ -120,14 +120,25 @@ public void init(VectorSchemaRoot historyRoot) {
...
+        _states.set(row, active ? 1 : 0);
+        _updates.set(row, updated ? 1 : 0);
+        _prevPtrs.set(row, prevPtr);
 
-        _edgeIds.setSafe(row, localEdgeId);
-        _isOutgoings.setSafe(row, outgoing ? 1: 0);
+        _edgeIds.set(row, localEdgeId);
+        _isOutgoings.set(row, outgoing ? 1: 0);
 
         //_snapshots.setSafe(row, snapshotRow);
```
Comment (miratepuffin) [inline arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexHistoryStore.java:143]: remove?

File: arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexPartition.java [hunk 10]
```diff
@@ -32,25 +32,29 @@
...
+    private boolean _loaded = false;        // Whether this partition has been loaded
+    private boolean _sorted = false;        // Whether this partition has been sorted
 
 
     // History, Fields and Properties:
-    protected final VertexHistoryPartition _history;           // The vertex history partition
-    //protected final VertexSnapshotPartition _snapshots;      // The vertex snapshots partition
-    protected final VertexEdgeIndexPartition _edgeIndex;       // Index of edges by dst-vertex-id
+    protected final VertexHistoryPartition _history;             // The vertex history partition
+    //protected final VertexSnapshotPartition _snapshots;        // The vertex snapshots partition
```
[the remark this answers is not in the document]
Comment (miratepuffin) [reply arrow-core/src/main/java/com/raphtory/arrowcore/implementation/VertexPartition.java:52]: Same with other snapshot comments below

File: core/project/build.properties [hunk 11]
```diff
@@ -1 +1 @@
-sbt.version=1.5.5
+sbt.version=1.8.2
```
Comment (miratepuffin) [inline core/project/build.properties:1]: Think we deleted this file in master. Do we need it?

**Review by miratepuffin — CHANGES_REQUESTED** [review]
Have finished having a read of arrow core - will do raphtory core when back from meeting. 

Main comments are:

- Quite a few blocks of code which are commented out, can it be checked if these are needed or if can be removed. If needed just a quick note on why so we remember.
- We have lots of commented out println's - any reason we can't use loggers.
- Otherwise looks great and the comments are 👌🏻

File: core/project/build.properties (hunk 11 above)
Comment (fabianmurariu) [reply core/project/build.properties:1]: we probably need to rebase

**Comment by fabianmurariu** [comment]
For some reason the tests run much much slower on this branch than in master

**Review by Alnaimi- — APPROVED** [review]

**Pull request merged**